src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/AbstractExpressionHandler.java

```diff
@@ -437,22 +437,23 @@ protected static DetailAST getFirstAstNode(DetailAST ast) {
      * @return the first ast of the expression
      */
     private static DetailAST getFirstAst(DetailAST ast, DetailAST tree) {
-        DetailAST realStart = ast;
-
-        if (tree.getLineNo() < realStart.getLineNo()
-            || tree.getLineNo() == realStart.getLineNo()
-            && tree.getColumnNo() < realStart.getColumnNo()
-        ) {
-            realStart = tree;
-        }
 
-        // check children
-        for (DetailAST node = tree.getFirstChild();
-            node != null;
-            node = node.getNextSibling()) {
-            realStart = getFirstAst(realStart, node);
+        DetailAST curNode = ast;
+        DetailAST realStart = ast;
+        DetailAST toVisit = tree;
+        while (curNode != null) {
```

Comment:
I think you may be going too far in the iterations. Original code looked at tree's position and it's children. It never went to `tree`'s parent which looks like it could be happening with your new version.

Reply:
@rnveach Fixed it now but pitest will still fail, test cases need to be added as a simple piece of code like - 
```
// This is WRONG
private static DetailAST getFirstAst(DetailAST ast) {

        DetailAST curNode = ast.getFirstChild();
        DetailAST realStart = ast;
        DetailAST toVisit = curNode;
        while (curNode != ast && curNode != null) {
            if (toVisit.getLineNo() < realStart.getLineNo()
                    || toVisit.getLineNo() == realStart.getLineNo()
                    && toVisit.getColumnNo() < realStart.getColumnNo()) {
                realStart = toVisit;
            }
            toVisit = curNode.getFirstChild();
            curNode = toVisit;
        }
        return realStart;
    }
```

which does not even traverse the whole tree passe all the test cases, some important cases are missing, lets see if we can find something in diff report.